controllers/etcdcluster_controller.go

```diff
@@ -83,14 +86,130 @@ func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 			log.Error(err, "unable to create Service", "namespace", service.Namespace, "cluster", cluster.Name)
 			return ctrl.Result{}, err
 		}
+		log.V(1).Info("Created Service", "service name", service.Name)
 		return ctrl.Result{}, nil
 	}
+	log.V(2).Info("Service exists", "service name", service.Name)
+
+	peers := &etcdv1alpha1.EtcdPeerList{}
+	if err := r.List(ctx, peers, client.MatchingFields{"spec.clusterName": cluster.Name}); err != nil {
+		log.Error(err, "unable to list peers", "namespace", service.Namespace, "cluster", cluster.Name)
+		return ctrl.Result{}, err
+	}
+
+	// Assume we can't communicate with the cluster
+	log.V(1).Info("Unable to communicate with etcd cluster")
+	if int32(len(peers.Items)) < *cluster.Spec.Replicas {
+
+		// Create more peers
+		peerName := nextAvailablePeerName(cluster, peers.Items)
+		log.V(1).Info("Insufficient peers for replicas, adding new peer",
+			"Current number of peers", len(peers.Items),
+			"Desired number of replicas", cluster.Spec.Replicas,
+			"New peer's name", peerName)
+		peer := peerForCluster(cluster, peerName)
+		if err := r.Create(ctx, peer); err != nil {
+			log.Error(err, "Failed to create peer",
+				"namespace", service.Namespace,
+				"cluster", cluster.Name,
+				"peer", peerName)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
 	return ctrl.Result{}, nil
 }
 
+func peerForCluster(cluster *etcdv1alpha1.EtcdCluster, peerName string) *etcdv1alpha1.EtcdPeer {
+	return &etcdv1alpha1.EtcdPeer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      peerName,
+			Namespace: cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(cluster, etcdv1alpha1.GroupVersion.WithKind("EtcdCluster")),
+			},
+			Labels: map[string]string{
+				appLabel:     appName,
+				clusterLabel: cluster.Name,
+			},
+		},
+		Spec: etcdv1alpha1.EtcdPeerSpec{
+			ClusterName: cluster.Name,
+			Bootstrap: &etcdv1alpha1.Bootstrap{
+				Static: &etcdv1alpha1.StaticBootstrap{
+					InitialCluster: initialClusterMembers(cluster),
+				},
+			},
+		},
+	}
+}
+
+func nthPeerName(cluster *etcdv1alpha1.EtcdCluster, i int) string {
+	return fmt.Sprintf("%s-%d", cluster.Name, i)
+}
+
+func initialClusterMembers(cluster *etcdv1alpha1.EtcdCluster) []etcdv1alpha1.InitialClusterMember {
+	names := expectedPeerNamesForCluster(cluster)
+	members := make([]etcdv1alpha1.InitialClusterMember, len(names))
+	for i, _ := range members {
+		members[i] = etcdv1alpha1.InitialClusterMember{
+			Name: names[i],
+			Host: expectedAdvertisePeerURLForPeer(cluster, names[i]),
+		}
+	}
+	return members
+}
+
+func nextAvailablePeerName(cluster *etcdv1alpha1.EtcdCluster, peers []etcdv1alpha1.EtcdPeer) string {
```

Comment:
how much do we care about the names being predictable? will it be required for the PVCs? if not, what's wrong with hashes?

Reply:
They have to be predictable, otherwise as you suggest we'll lose track of the PVCs.

Reply:
got it. i also have a question around "draining" a particular etcd node. maybe one of the nodes has some kind of memory leak or data corruption, and we want to replace it with a new one. will we
1) destroy the broken instance (and its PV) and recreate a new one with the same name

or

2) create a new instance (with a new, unique, name) and keep the old one around in a drained state

it'd prefer option 2) as throwing away broken things means you can't understand what broke it, but i don't think that would be supported with this peer naming scheme

Reply:
When you say "node" I assume you mean 'a member of the etcd cluster' and not 'a Kubernetes node running an etcd pod'?

Certainly as things stand it would be option one. We'd destroy it and it would get recreated.

We could allow, say, `my-cluster-2` to be marked as 'inactive' or somesuch, and allow `my-cluster-3` to be created as a new `EtcdPeer`, but then the peer controller would have to somehow block the pod for `my-cluster-3` from executing.

Alternatively, if a peer controller determined that one pod was corrupt, it could shut down the pod, rename the PV, and then recreate a fresh one. Leaving the old PV around for manual inspection?

Reply:
> Alternatively, if a peer controller determined that one pod was corrupt, it could shut down the pod, rename the PV, and then recreate a fresh one. Leaving the old PV around for manual inspection?

i don't think it's possible to rename PVs

can you elaborate on why we'd lose track of the PVCs if we don't have predictable names? it's my understanding that the peer controller will, given a named etcdpeer, create a replicaset and a pvc with that name. i.e the peer controller is responsible for keeping track of the PVC, and it wouldn't seem to matter what the name is?

Reply:
> I don't think it's possible to rename PVs

Ah. That would put that plan out of action then.

> it's my understanding that the peer controller will, given a named etcdpeer, create a replicaset and a pvc with that name. i.e the peer controller is responsible for keeping track of the PVC, and it wouldn't seem to matter what the name is?

Yes, so long as the `EtcdPeer` resource is not deleted. We were thinking of the case where the peer resource does get deleted (i.e., you delete everything except the `PersistantVolume`s) and if you made a replacement `EtcdCluster` each peer would have the same identity, connect to it's old data directory, and connect back together.

Reply:
let's continue talking about this on slack https://improbable.slack.com/archives/CNBN2EXFU/p1571392528153000

Reply:
resolution of this is that we understand the limitations of using sequential names, and will get them documented, but it's the simplest solution & will solve all the problems in front of us now so we'll stick to the naming as is.